AI revision workflow based on GPT-3 models (#484)
manubot/rootstock: USAGE.md

Words in `build/assets/custom-dictionary.txt` are ignored during spellchecking.
Spellchecking is currently only supported for English language manuscripts.

## AI-assisted authoring

> The workflow `ai-revision` is available to assist authors in writing their manuscripts.

Reviewer: I think this might work:

```suggestion
The workflow [`ai-revision`](.github/workflows/ai-revision.yaml) is available to assist authors in writing their manuscripts.
```

> It uses large language models (LLM) to revise the manuscript text, fixing spelling and grammar errors, and improving the sentence structure and the writing style with section-specific prompts.

Reviewer 2: ```suggestion
It uses large language models to revise the manuscript text, fixing spelling and grammar errors, and improving the sentence structure and the writing style with section-specific prompts.
```

It is manually triggered by the user (it never runs automatically), and it generates a pull request with suggested revisions.
Then the user can review these changes and merge the pull request if they are acceptable.

> More information about this tool is available in [this preprint](https://greenelab.github.io/manubot-gpt-manuscript/).

Reviewer: So we don't have to update once it's no longer a preprint:

```suggestion
More information about this tool is available in [this manuscript](https://greenelab.github.io/manubot-gpt-manuscript/).
```

You must provide a secret with name `OPENAI_API_KEY` containing your OpenAI API token.

> The easiest way to set this is to go to your organization's settings page and, within "Secrets and variables," select "Actions."

Reviewer 2: ```suggestion
The easiest way to set this is to go to your GitHub user or organization's settings page and, within "Secrets and variables," select "Actions."
```
This phrasing is awkward, but what is a better way to be inclusive or individual GitHub users as well?

> Next, create a secret with the appropriate name and the value of the API token.

Reviewer 2: ```suggestion
Next, create a secret with the name `OPENAI_API_KEY` and the value of the API token.
```

By default, the tool uses the model `text-davinci-003`.
Make sure to check the [pricing](https://openai.com/api/pricing/) of the OpenAI API.

> The cost for a revision of a standard manuscript should be around $0.50.

Reviewer 2: How many words is a "standard" manuscript?

Author: The median is 4,133 words, according to the [first Google result asking this](https://quantifyinghealth.com/length-of-a-research-paper/#:~:text=1%2D%20The%20median%20length%20of,between%202%2C023%20and%208%2C284%20words.). That's about 34 paragraphs, so it should be around that ($0.50). Our [manuscript about this tool](https://greenelab.github.io/manubot-gpt-manuscript/) had 34 paragraphs, and it cost $0.34.

I added this info in the sentence.

> The workflow allows to specify the branch and file names (in the `content/` directory) to revise, the language model to use, and the output branch name.

Reviewer 2: ```suggestion
The workflow allows specifying the branch and file names (in the `content/` directory) to revise, the language model to use, and the output branch name.
```

> Internally, the workflow uses the tool [Manubot AI Editor](https://github.com/greenelab/manubot-ai-editor) to revise the manuscript.

Reviewer 2: Noting the possibility of transferring to the manubot organization

Author: Updated.

For more advanced users, the behavior of the Manubot AI Editor or the parameters used for the language model can be changed using environment variables.
These variables can be changed in the workflow file (`ai-revision.yaml`).

> 

Reviewer 2: Per the prior discussion, this is where we could add a sentence or two about recommended practices for writing with AI tools. That could include a suggestion about listing them in the methods but not as co-authors.

Do we also want to add something here or in the "Citing Manubot" subsection about citing the GPT models and the Manubot AI Editor if they used for revision?

Author: I've added a few sentences about the controversy of using these tools in scholarly writing.

I leave the decision to you and @dhimmel about the citation.

Reviewer 2: If Manubot expands to become a suite of tools that are described in their on manuscripts, I like the idea of including them in the citation section. It encourages others to contribute their workflows or tools back to rootstock and/or the manubot package like you did.

What do you think @dhimmel?

Reviewer 2: @dhimmel we should make a decision about expanding the citation section before merging.

Reviewer: Yeah, I think it makes sense to add the ai revision citation to USAGE.

Reviewer: I don't think that we should demand that users of ai-revision subcommand cite the manuscript but can make them aware of the option.

Reviewer 2: Agreed, we don't need to demand citation. At the very end of "Citing Manubot" we can add something like

> To cite the Manubot AI Editor, see `@doi:10.1101/2023.01.21.525030`:
> 
> <reference using Manubot's default CSL>

@miltondp could you please add that?

Author: Done!

It is manually triggered by the user (it never runs automatically), and it generates a pull request with suggested revisions.
Then the user can review these changes and merge the pull request if they are acceptable.
More information about this tool is available in [this manuscript](https://greenelab.github.io/manubot-gpt-manuscript/).

You must provide a secret with name `OPENAI_API_KEY` containing your OpenAI API token.
The easiest way to set this is to go to your organization's settings page and, within "Secrets and variables," select "Actions."
Next, create a secret with the name `OPENAI_API_KEY` and the value of the API token.
By default, the tool uses the model `text-davinci-003`.
Make sure to check the [pricing](https://openai.com/api/pricing/) of the OpenAI API.
With $0.02 per 1000 tokens using the most powerful AI models, the cost for a revision of a standard manuscript (around 35 paragraphs) should be around $0.50.
The workflow allows specifying the branch and file names (in the `content/` directory) to revise, the language model to use, and the output branch name.

> Internally, the workflow uses the tool [Manubot AI Editor](https://github.com/greenelab/manubot-ai-editor) to revise the manuscript.

Reviewer 2: ```suggestion
Internally, the workflow uses the tool [Manubot AI Editor](https://github.com/manubot/manubot-ai-editor) to revise the manuscript.
```
Not sure what happened but it looks like this didn't actually update

For more advanced users, the behavior of the Manubot AI Editor or the parameters used for the language model can be changed using environment variables.
These variables can be changed in the workflow file (`ai-revision.yaml`).

It is important to note that using language models in scientific writing is a matter of debate among researchers and journal editors.
Therefore, it's advisable to follow the guidelines that journals and the research community propose.
For example, the *Nature* journal has published [rules about using language models in scholarly writing](https://www.nature.com/articles/d41586-023-00191-1), such as not listing the tools as authors and documenting how they were used.

> Since a Manubot-based manuscript uses GitHub, one approach consists in linking the AI-generated pull request, which will transparently show the changes suggested by the AI tool.

Reviewer 2: ```suggestion
Since a Manubot-based manuscript uses GitHub, one approach consists of linking the AI-generated pull request, which will transparently show the changes suggested by the AI tool.
```

I like your phrasing in this section. I don't want to add it to this section, but note that the new policy at Science prohibits using AI-generated text https://www.science.org/doi/10.1126/science.adg7879. It's not clear that AI-revised text would fall under that policy, and we shouldn't try to catalog all journals' policies here.

Author: We could add another sentence saying that authors should check the policy of the target journal before using this tool.

Reviewer 2: I'd say
> it's advisable to follow the guidelines that journals and the research community propose.

covers that well enough already

Spellchecking is currently only supported for English language manuscripts.

## AI-assisted authoring

The workflow [`ai-revision`](.github/workflows/ai-revision.yaml) is available to assist authors in writing their manuscripts.
It uses large language models to revise the manuscript text, fixing spelling and grammar errors, and improving the sentence structure and the writing style with section-specific prompts.
It is manually triggered by the user (it never runs automatically), and it generates a pull request with suggested revisions.
Then the user can review these changes and merge the pull request if they are acceptable.
More information about this tool is available in [this manuscript](https://greenelab.github.io/manubot-gpt-manuscript/).

You need to change your repository settings to 1) provide a secret with name `OPENAI_API_KEY` containing your OpenAI API token, and 2) allow workflows to create pull requests.
For 1), go to the settings page and, within "Secrets and variables," select "Actions."
Next, create a repository secret with the name `OPENAI_API_KEY` and the value of the API token (you can also do this using "Organization secrets" if available).

> For 2), go to "Actions", "General", "Workflow permissions", and activate the checkbox "Allow GitHub Actions to create and approve pull requests."

Author: I changed things a little bit when I added the second step to change the repo's settings.

By default, the tool uses the model `text-davinci-003`.
Make sure to check the [pricing](https://openai.com/api/pricing/) of the OpenAI API.

> With $0.02 per 1000 tokens using the most powerful AI models, the cost for a revision of a standard manuscript (around 35 paragraphs) should be around $0.50.

Reviewer: Wow! Much cheaper than a copy editor. That's wild.

The workflow allows specifying the branch and file names (in the `content/` directory) to revise, the language model to use, and the output branch name.
Internally, the workflow uses the tool [Manubot AI Editor](https://github.com/manubot/manubot-ai-editor) to revise the manuscript.
For more advanced users, the behavior of the Manubot AI Editor or the parameters used for the language model can be changed using environment variables.
These variables can be changed in the workflow file (`ai-revision.yaml`).

It is important to note that using language models in scientific writing is a matter of debate among researchers and journal editors.
Therefore, it's advisable to follow the guidelines that journals and the research community propose.
For example, the *Nature* journal has published [rules about using language models in scholarly writing](https://www.nature.com/articles/d41586-023-00191-1), such as not listing the tools as authors and documenting how they were used.

> Since a Manubot-based manuscript uses GitHub, one approach consists of linking the AI-generated pull request, which will transparently show the changes suggested by the AI tool.

Reviewer: 🔥
